cwl/src/main/scala/cwl/CommandOutputBinding.scala

```diff
@@ -1,15 +1,212 @@
+                             outputWomType: WomType,
+                             commandOutputBinding: CommandOutputBinding,
+                             secondaryFilesCoproduct: Option[SecondaryFiles],
+                             formatCoproduct: Option[StringOrExpression]): ErrorOr[WomValue] = {
+    val parameterContext = ParameterContext(inputs = inputValues)
+
+    // 3. outputEval: pass in the primary files to an expression to generate our return value
+    def evaluateWomValue(womFilesArray: WomArray): ErrorOr[WomValue] = {
+      commandOutputBinding.outputEval match {
+        case Some(StringOrExpression.String(string)) => WomString(string).valid
+        case Some(StringOrExpression.Expression(expression)) =>
+          val outputEvalParameterContext = parameterContext.copy(self = womFilesArray)
+          ExpressionEvaluator.eval(expression, outputEvalParameterContext)
+        case None =>
+          womFilesArray.valid
+      }
+    }
+
+    // Used to retrieve the file format to be injected into a file result.
+    def formatOptionErrorOr = CommandLineTool.CommandOutputParameter.format(formatCoproduct, parameterContext)
+
+    // 4. secondaryFiles: just before returning the value, fill in the secondary files on the return value
+    def populateSecondaryFiles(evaluatedWomValue: WomValue): ErrorOr[WomValue] = {
+      evaluatedWomValue match {
+        case womMaybePopulatedFile: WomMaybePopulatedFile =>
+          val secondaryFilesErrorOr = CommandLineTool.CommandOutputParameter.secondaryFiles(
+            womMaybePopulatedFile,
+            secondaryFilesCoproduct,
+            parameterContext
+          )
+
+          (secondaryFilesErrorOr, formatOptionErrorOr) mapN { (secondaryFiles, formatOption) =>
+            womMaybePopulatedFile.copy(secondaryFiles = secondaryFiles, formatOption = formatOption)
+          }
+
+        case womArray: WomArray if womArray.womType.memberType == WomMaybePopulatedFileType =>
+          formatOptionErrorOr map { formatOption =>
+            womArray.map(_.asInstanceOf[WomMaybePopulatedFile].copy(formatOption = formatOption))
+          }
+
+        case womValue: WomValue => womValue.valid
+      }
+    }
+
+    // CWL tells us the type this output is expected to be. Attempt to coerce the actual output into this type.
+    def coerceWomValue(populatedWomValue: WomValue): ErrorOr[WomValue] = {
+      (outputWomType, populatedWomValue) match {
+        case (womType: WomArrayType, womValue: WomArray) => womType.coerceRawValue(womValue).toErrorOr
+        case (womType: WomArrayType, womValue) =>
+          // Coerce a single value to an array
+          womType.coerceRawValue(womValue).toErrorOr.map(value => WomArray(womType, List(value)))
+        case (womType, womValue: WomArray) if womValue.value.lengthCompare(1) == 0 =>
+          // Coerce an array to a single value
+          womType.coerceRawValue(womValue.value.head).toErrorOr
```

Comment:
Not yet until CWL asks for this too. I don't believe `Array[File] files = <single file>` / `File single = <array files take head>` are valid WDL, while both are valid in CWL.

However if that changes this could go in Wom. I could also see a `womType.coerceRawValueHead` or something if we really wanted something like that down in Wom.